Add cop for `OS.linux?` and `OS.mac?` usage

apainintheneck opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

The methods `OS.linux?` and `OS.mac?` should only be used in `extend/os` and for tests and this cop makes sure of that.

I've created the cop first and I'll go about fixing the linter errors in other commits.

There will be a lot of small changes here so I'm going to split up the commits so that it's easier to review.

Closes #13967

dawidd6 commented on Library/Homebrew/rubocops/platform.rb:
```diff
@@ -0,0 +1,25 @@
+      #
+      # @api private
+      class MoveToExtendOS < Base
+        MSG = "Move calls to `OS.linux?` and `OS.mac?` to `extend/os`."
```
```suggestion
        MSG = "Move `OS.linux?` and `OS.mac?` calls to `extend/os`."
```

Sounds better this way. At least for me 😄

apainintheneck replied:
Yeah, that does sound better. I'll wait until the end to change it just in case there are any other naming changes people want.

apainintheneck replied:
Added in [aff4f27](https://github.com/Homebrew/brew/pull/14163/commits/aff4f271c39bac51a6b5822984766752bd1edece).

apainintheneck commented:
Does it make sense to not run this cop against requirements?

```sh
Library/Homebrew/requirements/linux_requirement.rb:12:31: C: Move calls to OS.linux? and OS.mac? to extend/os.
  satisfy(build_env: false) { OS.linux? }
                              ^^^^^^^^^
Library/Homebrew/requirements/macos_requirement.rb:54:5: C: Move calls to OS.linux? and OS.mac? to extend/os.
    OS.mac? && @version
    ^^^^^^^
Library/Homebrew/requirements/macos_requirement.rb:59:18: C: Move calls to OS.linux? and OS.mac? to extend/os.
    next true if OS.mac?
                 ^^^^^^^
```

apainintheneck commented:
@carlocab The following snippet exists in two places in `Library/Homebrew/formula_cellar_checks.rb`. Is there another way to check for the generic OS that doesn't use these methods?

```rb
# There is no `binary_executable_or_library_files` method for the generic OS
return if !OS.mac? && !OS.linux?
```

apainintheneck commented:
```
simulate_system.rb:16: Expected T::Boolean but found NilClass for method result type https://srb.help/7005
    16 |      def simulating_or_running_on_macos?; end
                                                   ^^^
  Expected T::Boolean for result type of method simulating_or_running_on_macos?:
    simulate_system.rb:16:
    16 |      def simulating_or_running_on_macos?; end
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Got NilClass originating from:
    simulate_system.rb:16: Possibly uninitialized (NilClass) in:
    16 |      def simulating_or_running_on_macos?; end
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Errors: 1
```

I wanted to leave empty method definitions as documentation in classes but it seems like Sorbet isn't always a fan of that. I think this is because other files use these methods and so Sorbet expects a certain return type that isn't nil. Is there an elegant way to handle this other than just removing those methods and only including them in `extend/os`?

apainintheneck commented:
> IMO the extend system feels weird for `(dev-)cmd` given it's largely API-less, but could make sense of things extracted as proper API methods that live outside of commands like upgrade.rb.

Yeah, it is kind of awkward. We could of course just not run the cop on those directories. They did require a bit more refactoring than other areas of the codebase since they're often smack dab in the middle of a giant method.

apainintheneck commented:
Okay, I fixed a few more linter errors, changed rubocop message as suggested by @dawidd6, and threw the last few linter errors into the `.rubocop_todo.yml` file.

MikeMcQuaid commented:
> IMO the extend system feels weird for `(dev-)cmd` given it's largely API-less, but could make sense of things extracted as proper API methods that live outside of commands like upgrade.rb.

Yeh, in general the code in `(dev-)cmd` should be super minimal (but often isn't) and if this cop nudges more code being extracted to proper methods: all the better.

> Sounds like the default method should be the generic os version.

Yup 👍🏻.

> It does make me wonder though how much we should worry about code duplication. It seems like adding some things to `extend/os` actually increases the number of lines of code.

That's fine for me. It's not more logic really, just more boilerplate.



> I'm open to this especially if it'll make it easier for you all to review.

Yes, definitely would. Would rather land three smaller PRs more quickly than one bigger, personally.



> The problem with this is that we'll end up with duplicate code in `extend/os` since the Linux and Mac versions of the function are identical.

Ok, stupid question: why does the generic version have to be different then?

MikeMcQuaid reviewed (approved):
Looks good!

apainintheneck commented:
> Ok, stupid question: why does the generic version have to be different then?

It probably doesn't have to be to be honest. Looks like an optimization.

carlocab commented:
> @carlocab The following snippet exists in two places in `Library/Homebrew/formula_cellar_checks.rb`. Is there another way to check for the generic OS that doesn't use these methods?
> 
> ```ruby
> # There is no `binary_executable_or_library_files` method for the generic OS
> return if !OS.mac? && !OS.linux?
> ```
> 
> Edit: We could also just add this to the OS module under `OS.generic?`. Or we could create a separate section for the generic OS in `extend/os`.

I _think_ something like this will work (warning: untested)
```diff
diff --git a/Library/Homebrew/extend/os/keg.rb b/Library/Homebrew/extend/os/keg.rb
index 5e3cdd073..94798c864 100644
--- a/Library/Homebrew/extend/os/keg.rb
+++ b/Library/Homebrew/extend/os/keg.rb
@@ -1,4 +1,8 @@
 # typed: strict
 # frozen_string_literal: true
 
-require "extend/os/mac/keg" if OS.mac?
+if OS.mac?
+  require "extend/os/mac/keg"
+elsif
+  require "extend/os/linux/keg"
+end
diff --git a/Library/Homebrew/extend/os/linux/keg.rb b/Library/Homebrew/extend/os/linux/keg.rb
new file mode 100644
index 000000000..9b0eaf81d
--- /dev/null
+++ b/Library/Homebrew/extend/os/linux/keg.rb
@@ -0,0 +1,10 @@
+# typed: true
+# frozen_string_literal: true
+
+class Keg
+  undef binary_executable_or_library_files
+
+  def binary_executable_or_library_files
+    elf_files
+  end
+end
diff --git a/Library/Homebrew/keg.rb b/Library/Homebrew/keg.rb
index a12e8de39..052d433d2 100644
--- a/Library/Homebrew/keg.rb
+++ b/Library/Homebrew/keg.rb
@@ -527,9 +527,7 @@ class Keg
     find { |pn| FileUtils.rm_rf pn if pn.basename.to_s == "__pycache__" }
   end
 
-  def binary_executable_or_library_files
-    elf_files
-  end
+  def binary_executable_or_library_files; end
 
   def codesign_patched_binary(file); end
```

carlocab commented:
That said, I feel like calling `binary_executable_or_library_files` should actually throw a  "not implemented" error when called in the generic OS, if the idea is to make porting easier. But that will turn into a whole new yak shave.

Bo98 commented:
> Yeh, in general the code in `(dev-)cmd` should be super minimal (but often isn't) and if this cop nudges more code being extracted to proper methods: all the better.

I mean sure, I agree if that was the route that was actually taken, but that does not seems to be the direction taken here. The direction appears to have been to keep it there and plop it in `extend/cmd` which I don't agree is the best direction.

apainintheneck commented:
> The direction appears to have been to keep it there and plop it in extend/cmd which I don't agree is the best direction and I think makes it worse to read really given it's knee deep in file-private implementation methods. Was also hoping the assumption that command stuff stays in `(dev-)cmd` would stick so we could add `--exclude (dev-)cmd` to `.yardopts` and block anything from `require`-ing those files outside of the CLI runner.

I didn't know this was a thing. Sounds like I should probably just move the offending `cmd/` and `dev-cmd/` files to `.rubocop_todo.yml` since it's probably out of the scope of this PR and my current approach goes against everything you just mentioned.

apainintheneck commented:
Okay, I removed all of the `cmd/` and `dev-cmd/` related commits so I think this will be ready once tests pass (waiting on resolution of #14175).

MikeMcQuaid commented on Library/Homebrew/.rubocop_todo.yml:
```diff
@@ -36,3 +36,13 @@ Lint/EmptyBlock:
     - "test/pathname_spec.rb"
     - "test/support/fixtures/cask/Casks/*flight*.rb"
+
+Homebrew/MoveToExtendOS:
+  Exclude:
+    - "cask/cmd/install.rb"
+    - "cmd/update-report.rb"
+    - "default_prefix.rb"
+    - "dev-cmd/bottle.rb"
+    - "dev-cmd/tests.rb"
+    - "dev-cmd/update-test.rb"
+    - "formula_cellar_checks.rb"
```
Sorry for all the back and forth: what do you think about individual `#rubocop:disable` comments for these explaining each one and adding an inline `TODO` to go clean/fix them up? I think that's probably more likely to end up fixed than in here.

apainintheneck replied:
The other advantage is that it will trigger a linter error if people try to add more of those OS checks to those files in the future. I like the idea.

MikeMcQuaid commented:
@apainintheneck the `brew doctor` issue is resolved but I think I screwed up the merge conflict resolution here, sorry!

apainintheneck commented:
> @apainintheneck the `brew doctor` issue is resolved but I think I screwed up the merge conflict resolution here, sorry!

No worries. Thanks for giving me a heads up! Hopefully, it will work after this final rebase.

MikeMcQuaid commented:
Thanks again @apainintheneck!

Bo98 commented:
Is your formula perhaps not in a tap?

johnrengelman commented:
Pretty new to these parts of Homebrew, so sorry if I'm doing some totally wrong.
I'm trying to automate CI for a private TAP in my organization.
So I have the TAP repository and my CI is checking out the repository and then running `brew audit --formula` on each file under `Formulas/`. 
So I'm not doing anything to "install the tap" as part of that. Should I be?

apainintheneck commented:
I took another look at the `.rubocop.yml` file and found that we ignore files in `Formula/` and `Cask/` folders when linting other rules we don't want to apply to formulae and casks. I'll add that to the list and then it'd be great if you could test it out.

apainintheneck commented:
@johnrengelman Can you try it again with the changes on this branch https://github.com/Homebrew/brew/pull/14236?
